src/coreclr/jit/scev.cpp

```diff
@@ -532,6 +574,10 @@ Scev* ScalarEvolutionContext::AnalyzeNew(BasicBlock* block, GenTree* tree, int d
                 case GT_ADD:
                     oper = ScevOper::Add;
                     break;
+                case GT_SUB:
```

Comment:
@jakobbotsch, just curious, could we also handle `GT_MOD` here? e.g. native compilers produce https://godbolt.org/z/a38Ms6GW, while net9.0 daily build (9.0.100-preview.3.24151.2) gives:

<details>
  <summary>linux arm64</summary>

```asm
  ; Assembly listing for method nine.C:Foo(int):int:this (FullOpts)
  ; Emitting BLENDED_CODE for generic ARM64 - Unix
  ; FullOpts code
  ; NativeAOT compilation
  ; optimized code
  ; fp based frame
  ; fully interruptible

  G_M000_IG01:                ;; offset=0x0000
              stp     fp, lr, [sp, #-0x10]!
              mov     fp, sp

  G_M000_IG02:                ;; offset=0x0008
              mov     w0, wzr
              mov     w2, wzr
              cmp     w1, #0
              ble     G_M000_IG06
              align   [4 bytes for IG03]
              align   [4 bytes]
              align   [0 bytes]
              align   [0 bytes]
   
  G_M000_IG03:                ;; offset=0x0020
              tbz     w2, #0, G_M000_IG05
   
  G_M000_IG04:                ;; offset=0x0024
              add     w0, w0, w2
   
  G_M000_IG05:                ;; offset=0x0028
              add     w2, w2, #2
              cmp     w2, w1
              blt     G_M000_IG03
   
  G_M000_IG06:                ;; offset=0x0034
              ldp     fp, lr, [sp], #0x10
              ret     lr
   
  ; Total bytes of code 60
```
</details>

<details>
  <summary>osx arm64</summary>
```
  ; Assembly listing for method Program:<<Main>$>g__Foo|0_0(int):int (FullOpts)
  ; Emitting BLENDED_CODE for generic ARM64 - Apple
  ; FullOpts code
  ; NativeAOT compilation
  ; optimized code
  ; fp based frame
  ; fully interruptible
  
  G_M000_IG01:                ;; offset=0x0000
              stp     fp, lr, [sp, #-0x10]!
              mov     fp, sp
   
  G_M000_IG02:                ;; offset=0x0008
              mov     w1, wzr
              mov     w2, wzr
              cmp     w0, #0
              ble     G_M000_IG06
              align   [4 bytes for IG03]
              align   [4 bytes]
              align   [0 bytes]
              align   [0 bytes]
   
  G_M000_IG03:                ;; offset=0x0020
              tbz     w2, #0, G_M000_IG05
   
  G_M000_IG04:                ;; offset=0x0024
              add     w1, w1, w2
   
  G_M000_IG05:                ;; offset=0x0028
              add     w2, w2, #2
              cmp     w2, w0
              blt     G_M000_IG03
   
  G_M000_IG06:                ;; offset=0x0034
              mov     w0, w1
   
  G_M000_IG07:                ;; offset=0x0038
              ldp     fp, lr, [sp], #0x10
              ret     lr
   
  ; Total bytes of code 64
```
</details>

Reply:
It would definitely be representable, and the folding rules are simple (`<L, start, step> % x => <L, start % x, step % x> % x` which reduces to `start % x` if `x` divides `step`). We don't have a good fit for where we would actually do a simplification on the IR like that today given where scalar evolution analysis runs. It raises an interesting question about whether we should try to do the scalar evolution analysis as part of value numbering instead.

Reply:
> It raises an interesting question about whether we should try to do the scalar evolution analysis as part of value numbering instead.

I was tempted to suggest this too, but working with "trees" of VNFuncs is currently pretty clunky. I have thought about introducing a vistor pattern here but that never got past the thinking stage. Also some VNFuncs do odd things with their arguments.

As we get deeper into loop opts, we'll want to do more materialization of the SCEVs as IR, and VN for that is also clunky, since it doesn't track which available operands have the VNs you need. So I think the SCEvs need to be more IR like but perhaps keep tabs on the associated VNs too.

Reply:
Yeah, it might not be appropriate to switch it entirely to happen inside VN. Perhaps VN can just reinvoke the SCEV analysis that's there for primary IVs... that doesn't seem to be so expensive, and adding `VNF_AddRec` and the additional folding rules around it shouldn't be so complicated.